[AArch64] Implement -fno-plt for SelectionDAG/GlobalISel

MaskRay opened the pull request:
Clang sets the nonlazybind attribute for certain ObjC features. The
AArch64 SelectionDAG implementation for non-intrinsic calls
(46e36f0953aabb5e5cd00ed8d296d60f9f71b424) is behind a cl option.

GCC implements -fno-plt for a few ELF targets. In Clang, -fno-plt also
sets the nonlazybind attribute. For SelectionDAG, make the cl option not
affect ELF so that non-intrinsic calls to a dso_preemptable function use
GOT. Adjust AArch64TargetLowering::LowerCall to handle intrinsic calls.

For FastISel, change `fastLowerCall` to bail out when a call is due to
-fno-plt.

For GlobalISel, handle non-intrinsic calls in CallLowering::lowerCall
and intrinsic calls in AArch64CallLowering::lowerCall (where the
target-independent CallLowering::lowerCall is not called).
The GlobalISel test in `call-rv-marker.ll` is therefore updated.

Note: the current -fno-plt -fpic implementation does not use GOT for a
preemptable function.

Link: #78275

MaskRay committed:
[𝘀𝗽𝗿] initial version

Created using spr 1.3.4

davemgreen commented:
It looks like some of the tests might be failing? Or does it need a rebase?

arsenm commented on llvm/lib/CodeGen/GlobalISel/CallLowering.cpp:
```diff
@@ -144,9 +144,16 @@ bool CallLowering::lowerCall(MachineIRBuilder &MIRBuilder, const CallBase &CB,
-  if (const Function *F = dyn_cast<Function>(CalleeV))
-    Info.Callee = MachineOperand::CreateGA(F, 0);
-  else if (isa<GlobalIFunc>(CalleeV) || isa<GlobalAlias>(CalleeV)) {
+  if (const Function *F = dyn_cast<Function>(CalleeV)) {
+    if (F->hasFnAttribute(Attribute::NonLazyBind)) {
```
Should this check the CB attributes instead of directly looking at the underlying declaration's?

MaskRay replied:
`nonlazybind` is not used as call site attributes.

MaskRay committed:
fix FastISel failures by not classifying Addr.getGlobalValue(), which can be null

Created using spr 1.3.4

MaskRay commented:
> It looks like some of the tests might be failing? Or does it need a rebase?

Sorry about it. The last minute `FastISel` change caused some failures. `Addr.getGlobalValue()` could be null. I have simplified `AArch64FastISel.cpp` to just check `"RtLibUseGOT"` (added by `-fno-plt`) and fixed the failures.

arsenm commented on llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp:
```diff
@@ -1293,8 +1293,19 @@ bool AArch64CallLowering::lowerCall(MachineIRBuilder &MIRBuilder,
+    if (Info.Callee.isSymbol() && F.getParent()->getRtLibUseGOT()) {
+      auto Reg =
+          MRI.createGenericVirtualRegister(getLLTForType(*F.getType(), DL));
+      auto MIB = MIRBuilder.buildInstr(TargetOpcode::G_GLOBAL_VALUE);
+      DstOp(Reg).addDefToMIB(MRI, MIB);
```
use buildGlobalValue

MaskRay replied:
This part is to handle library calls.
`buildGlobalValue` builds a `G_GLOBAL_VALUE` with `MIB.addGlobalAddress(GV);`. Here we use an ExternalSymbol, and cannot use `buildGlobalValue`

arsenm replied:
Missing overload then, should try to avoid raw buildInstr calls when possible

MaskRay replied:
Did you propose a new MachineIRBuilder API? This is a special use of `buildInstr(TargetOpcode::G_GLOBAL_VALUE)` and there is only one. I think avoid adding another API is good for now.

MaskRay replied:
`buildInstr(TargetOpcode::G_GLOBAL_VALUE)` cannot be CSEed today, and I think this patch should not change it.

It seems that only the `buildInstr` overload with `DstOps/SrcOps` can perform CSE. `MachineIRBuilder::buildGlobalValue` does not call this overload.

arsenm replied:
Using raw buildInstr is a worse API. Please just add the new overload

MaskRay replied:
I disagree. I think adding a function similar to `buildGlobalValue`, either omits `const GlobalValue *GV` or changes the argument to `const char*`, will add confusion, as this is a very special use case.

I slightly simplified the code in the just-pushed commit.

davemgreen commented on llvm/test/CodeGen/AArch64/call-rv-marker.ll:
```diff
@@ -201,17 +201,27 @@ define dso_local void @rv_marker_3() personality ptr @__gxx_personality_v0 {
+; GISEL-NEXT:    ldp x29, x30, [sp, #16] ; 16-byte Folded Reload
 ; GISEL-NEXT:    ldp x20, x19, [sp], #32 ; 16-byte Folded Reload
-; GISEL-NEXT:    b _objc_release
+; GISEL-NEXT:    br x1
```
@fhahn, @TNorthover do these sound OK to you?

MaskRay replied:
Ping:) @fhahn @TNorthover

MaskRay committed:
simplify intrinsic call in AArch64CallLowering.cpp

Created using spr 1.3.4

MaskRay commented:
Ping:)

arsenm commented on llvm/lib/CodeGen/GlobalISel/CallLowering.cpp:
```diff
@@ -144,9 +144,16 @@ bool CallLowering::lowerCall(MachineIRBuilder &MIRBuilder, const CallBase &CB,
-  else if (isa<GlobalIFunc>(CalleeV) || isa<GlobalAlias>(CalleeV)) {
+  if (const Function *F = dyn_cast<Function>(CalleeV)) {
+    if (F->hasFnAttribute(Attribute::NonLazyBind)) {
+      auto Reg =
+          MRI.createGenericVirtualRegister(getLLTForType(*F->getType(), DL));
+      MIRBuilder.buildGlobalValue(Reg, F);
```
```suggestion
      Register Reg = MIRBuilder.buildGlobalValue(getLLTForType(*F->getType(), DL), F).getReg(0);
```

MaskRay replied:
After clang-format, this becomes
```
-      auto Reg =
-          MRI.createGenericVirtualRegister(getLLTForType(*F->getType(), DL));
-      MIRBuilder.buildGlobalValue(Reg, F);
+      Register Reg =
+          MIRBuilder.buildGlobalValue(getLLTForType(*F->getType(), DL), F)
+              .getReg(0);
```

which is not shorter... In addition, the new code `AArch64CallLowering.cpp` uses `createGenericVirtualRegister`. So sticking with `createGenericVirtualRegister` here adds consistency.

MaskRay replied:
Done in de93d8643072ee080aa26ecd9e952500cb39ab00 !

MaskRay committed:
replace MRI.createGenericVirtualRegister with buildGlobalValue

Created using spr 1.3.6-beta.1

davemgreen reviewed (approved):
From what I understand from an AArch64 perspective I think this looks OK too. Thanks

MaskRay committed:
replace MRI.createGenericVirtualRegister in AArch64CallLowering.cpp

Created using spr 1.3.6-beta.1

mstorsjo commented:
This change had a surprising effect on how dllimport attributes work for Windows targets:
```c
void __declspec(dllimport) foo(void);
void myFunc(void) { foo(); }
```
Before:
```
$ bin/clang -target aarch64-windows -S -o - dllimp.c -O2 -fno-plt
myFunc:
	adrp	x0, __imp_foo
	ldr	x0, [x0, :lo12:__imp_foo]
	br	x0
```
After:
```
$ bin/clang -target aarch64-windows -S -o - dllimp.c -O2 -fno-plt
myFunc:
	adrp	x0, foo
	ldr	x0, [x0, :lo12:foo]
	br	x0
```
(Obviously, this change in the output code breaks it entirely.)

I'm not entirely sure if this should be considered a bug or not - I don't think `-fno-plt` is supposed to have any effect on Windows targets - but it's somewhat unexpected that it makes the compiler produce broken code.
